### ISSUE TITLE
Mozart should provide back pressure in drawing pipeline

### ISSUE THREAD

**Issue by abarth** [domokit/mojo#684]
Currently, we just produce frames at the vsync frequence regardless of whether those frames get dropped later in the pipeline.

See https://github.com/flutter/flutter/issues/860

**Comment by krisgiesing** [comment]
Can you elaborate a bit on what causes the frames to get dropped later in the pipeline?

**Comment by abarth** [comment]
The underlying cause is that it's easy to quickly generate a lots of graphics commands that take the GPU a long time to process.  Currently, we produce graphics commands on a regular interval according to the screen's refresh rate regardless of how long it take the GPU to process the commands.

In the guts of Mozart, there's a frame queue, a lock, and a constant that limits how deep the queue is permitted to get.  When we stuff too many things into the queue and exceed the constant, Mozart drops frames from the queue:

https://github.com/domokit/mojo/blob/master/services/gfx/compositor/backend/gpu_output.h#L44

It's a waste of resources to generate frames that will get dropped.  Instead, there should be some signal from the end of the pipeline all the way to the beginning of the pipeline that we're over budget and to adapt accordingly.

How exactly to structure this backpressure signal is quite delicate, as is how to adapt earlier in the pipeline when receiving the signal.

**Comment by j9brown** [comment]
There are a bunch of things which make solving this problem hard today:

1. Apps can easily overwhelm the GPU process.
2. There's no accounting for the amount of outstanding work.
3. There's no way to prioritize processing to ensure that cycles are spent in the places where they will have the greatest impact.
4. The compositor's rendering time budget is unknown due to the aforementioned factors.
5. The compositor's scene graph representation is just rich enough that it runs the risk of becoming a dumping ground for unbounded application requested work.  We might need to impose scene complexity limits at some point.
6. Everything is asynchronous.
6. Everything is asynchronous.
6. Everything is asynchronous.

SurfaceFlinger had it easy...

In general I'm leaning towards the idea of an isochronous compositor output stage.  Each output interval, the compositor collects a snapshot of what it needs to draw making sure that all of the dependencies have been satisfied (all fences cleared), performs composition, sends the pixels to the hardware, then waits for the next interval.

It's kind of brute force but it should keep the compositor's critical path clear of obstacles.  Once it's decided what to draw, it can just rip through that.  Bonus points if we can prioritize its work somehow to give it a predictable rendering budget.  Basically I want this loop to be able to free-run at 60 Hz (120 Hz?) for as long as there is work to do.  (It can still overlap CPU and GPU work but there will never be more than 1 frame outstanding.)

Having thus isolated the output stage, the queuing then ends up happening closer to the producer side at the point where the producer is submitting new scenes.  To apply backpressure, the compositor can take note of how long the producer's fences take to clear and adjust the latency to compensate.

Example:

Suppose there are two apps, Rabbit and Sloth.

When Rabbit submits scene updates to the compositor, they're always immediately ready to go.  The compositor never has to wait at all for sync points to be signaled.  The compositor can always snapshot Rabbit's drawing commands and texture handles then submit them for rasterization on the next output interval.  Rabbit is fast so it can operate with lower presentation latency and virtually no queueing.

When Sloth submits scene updates to the compositor, they are frequently pending a bunch of GPU work.  The compositor notices that this work is almost never done in time to be snapshotted and rasterized so it usually has to wait an extra frame for Sloth's resources to be usable but it consistently gets there.  The compositor increases Sloth's presentation latency by 1 frame to account for the pipelining.

Meet Alligator.  Alligator is prone to spurts of activity followed by torpor.  The compositor hates Alligator because Alligator often tries to cram a whole bunch of work in too short an interval to keep up.  As a punishment, the compositor caps Alligator's presentation latency at 2 frames and stops delivering "update" signals until the work settles down again.  When Alligator happens to be in a listening mood, it will slow down.  But sometimes Alligator's ears are full of water and it races ahead regardless.  When this happens, the compositor has no choice but to discard some of Alligator's work.

Note 1: When an app submits a frame to Mozart, it's expected to supply a "presentation time stamp".  This value should come from the scheduler's FrameInfo (or may be defaulted).  The presentation latency is increased by specifying time stamps which are further in the future when the update is submitted.  This has the result of causing updates to be queued for longer thus effectively resulting in control over the pipeline depth.  Choosing the appropriate latency is left up to the producer and the compositor to negotiate.

(Having written this, I wonder if I should add an API to let the app specify its minimum latency to the scheduler so the deadline can be tuned more naturally.  It would be a little bit like tweaking double/triple buffering modes.)

Note 2: Scheduling is performed based on "presentation time stamp" rather than "frames" to simplify audio/video synchronization.  Moreover "updates" don't represent "frames" per se, they're just mutations of a scene graph which get coalesced at presentation time.  That's why you won't find a literal frame queue anywhere.

Note 3: The compositor may not need to way for a full snapshot to occur before it begins precomputing certain information it will need for rendering.  For example, it could start tesselating paths to generate shadow volumes as soon as those resources are updated.  TBD what will be worth pipelining here.

(Now I probably need to go write a design doc with pretty pictures illustrating how this all works out as a dynamic system...)

**Comment by abarth** [comment]
That sounds about right, at least for latency.  You might also want to consider throughput.  For some producers, you might want them to produce half as many updates in order to meet the overall throughput budget of the system.  That's not quite the Alligator case because Alligator is inconsistent whereas this producer (I'll let you select the appropriate animal avatar) is consistently supply fragment shaders that are too complicated (or whatever they're doing to blow the throughput budget).

**Comment by j9brown** [comment]
Throughput is challenging.  We may need some kind of accounting to identify hot spots, policies around prioritization or quota to control when work is issued, or some form of preemption of bad actors in addition to a feedback mechanism.

I vote for Snakes because they are sneaky and slippery.

**Comment by j9brown** [comment]
A few more thoughts on this.

1. Tracking scene updates doesn't tell the compositor anything about the GPU backlog.

The compositor knows when SceneUpdates are been published and how many are outstanding.  However each scene update doesn't necessarily represent a frame; it's just a mutation of the scene graph.  It's ok to repeatedly submit updates at any time and it can be quite cheap if it's only updating nodes or metadata.  At some point the compositor will snapshot the frame and that's the content we'll see.

In other words, the number of outstanding scene updates is irrelevant for applying backpressure.

2. Tracking frame retirement only tells the compositor about GPU backlog for on-screen content.

The compositor only inserts sync points for textures it needs to show on-screen.  If a texture is not used in a frame then the frame may be retired before the texture is produced.  If this happens often, the producer of these textures may receive signals to keep producing textures even when far backlogged.

In other words, the compositor is unable to apply backpressure on unseen work.  (Perhaps it should implicitly wait on all published textures even when unused?  Arguably clients are responsible for tracking their own backlog but it's easy to get it wrong.)

3. Feed-forward design fundamentally makes it hard to ensure all work is productive.

The compositor tells its clients when to publish scene updates but it doesn't tell them anything what they need to produce.  In particular, it deals poorly with occlusion or with unattached content.  Clients can go to a lot of trouble to produce scene updates and rasterize textures which are unused.  It's difficult to handle content of indefinite size.


Maybe we should flip it around and make it demand-based.  In such a model, the compositor would ask each producer of a scene to generate updated content at the beginning of each frame.  Granted this makes for a deep pipeline.

Or we could consider a checkpointing model where updates are mostly feed-forward but there's a mechanism for propagating important parameters regarding the demand for scene updates (eg. what portion of the scene is visible).

... but if we follow this thinking too far we end up in a world where the compositor has to drive animations and everything too which is something we're trying to avoid.

**Comment by j9brown** [comment]
Ok here's how it's going to work.

* Compositor stays feed-forward and remains responsible for delivering scene update pulses (on vsync) to all interested scenes simultaneously.

* Compositor will determine for each scene which renderer will drive that scene's scheduling.  A tie-breaking rule will be applied for scenes used in multiple renderer scene graphs.  Scenes which are unrendered will be associated with a default null renderer (possibly clocked at a very low rate to make forward progress or stalled).

* Compositor will insert waits on texture sync points at point of use in each rendered frame.

* Compositor will insert waits on all published but unused textures from scenes associated with the renderer after each rendered frame.  This is to ensure that the renderer can apply backpressure on the scenes that it is responsible for scheduling even when those scenes are not currently visible.

* Compositor will keep track of number of outstanding frames in the rendering pipeline and adjust scheduling when backlogged.  As a simple expedient, it may suppress updates until the number of outstanding frames drops below a set threshold.  Later we could consider adapting the update frequency itself.  (see also variable vsync displays)

* Compositor will not be responsible for propagating visibility / occlusion data to scenes.  Instead, handle this via the View Manager layout protocol.  For example, when hosting a view of indefinite size, a parent view would be responsible for setting view properties to indicate what portion of the child view that the child should render.  The child view should then update its scene accordingly.

* View Manager will apply a recursive invalidation mechanism for propagating view properties down the view tree.  This happens independently (and concurrently) with scene updates and relies on the Compositor's scene versioning scheme for optimistic synchronization.

* Clients are responsible for applying their own backpressure for any off-screen rendering.  On-screen drawing can be gated by the Compositor's scheduled scene updates given the changes to the renderer sync point handling listed above.

In a nutshell, the limitations of the Compositor feed-forward design will be compensated for by the View Manager's recursive view property propagation mechanism.

**Comment by krisgiesing** [comment]
Is there a mechanism for associating the timing of compositor scene update pulses with View Manager property change propagation? I'm wondering about side effects like tearing if views are trying to coordinate animations.

**Comment by j9brown** [comment]
Adam and I talked this out and circled back to an earlier sketch.  So we'll
do this instead:

* Views request an update pulse by sending an Invalidate() message.

* View Manager issues an update pulse to views providing a structure
containing view property updates and information about the frame to be
produced and which is expected to reply with a structure containing desired
changes to the contents of the view such as resizing children.  Updates are
propagated recursively down the view tree as required to satisfy
invalidated views.

* There would thus be two mechanisms to time updates.  Views should listen
to the update pulse from the View Manager and use the invalidation
mechanism to request updates.  Lower level things like the View Manager
itself may listen to the SceneScheduler from the Compositor instead.

**Comment by j9brown** [comment]
(This stuff was going to happen anyhow but originally I thought apps would
listen to the view update pulse for layout and scene update pulse for
drawing.  In the end it's easier and likely more efficient to provide a
single coupled event to work with.)

**Comment by jamesr** [comment]
> Compositor will insert waits on texture sync points at point of use in each rendered frame.

It'll be more efficient to insert waits for all textures that a scene needs before issuing the compositor's draw commands for the scene.  Otherwise the compositor's context may become blocked in the middle of its draw stream multiple times to run commands from other contexts to satisfy the wait which will be more expensive than flushing all of the dependencies before the commands start.  (This is true whether the state tracking is done by the command buffer service using context virtualization or by the underlying GL driver)
